services/api/service.go

```diff
@@ -830,6 +888,50 @@ func (api *RelayAPI) handleGetPayload(w http.ResponseWriter, req *http.Request)
 		if err != nil {
 			log.WithError(err).Error("failed to increment builder-stats after getPayload")
 		}
+
+		builderStatus, err := api.redis.GetBlockBuilderStatus(bidTrace.BuilderPubkey.String())
+		if err != nil {
+			log.WithError(err).Error("failed to get block builder status")
+		}
+
+		// Check if the block was valid in the optimistic case.
+		if builderStatus == common.Optimistic {
+			simErr := api.simulateBlock(blockSimOptions{
+				ctx:      req.Context(),
+				log:      log,
+				highPrio: true, // manually set to true for these blocks.
+				req: &BuilderBlockValidationRequest{
+					BuilderSubmitBlockRequest: types.BuilderSubmitBlockRequest{
+						Signature:        payload.Signature,
+						Message:          &bidTrace.BidTrace,
+						ExecutionPayload: getPayloadResp.Data,
+					},
+				},
+			})
+			if simErr != nil {
+				log.WithError(err).Error("failed to simulate signed block")
+				err = api.db.SaveValidatorRefund(bidTrace, payload)
+				if err != nil {
+					log.WithError(err).WithFields(logrus.Fields{
+						"bidTrace":                 bidTrace,
+						"signedBlindedBeaconBlock": payload,
+					}).Error("failed to save validator refund to database")
+				}
+
+				// Validation failed, mark the status of the builder as lowPrio (pessemistic).
+				newStatus := datastore.MakeBlockBuilderStatus(common.LowPrio)
```

Comment:
If simulation fails, would it be fair to say that demotion to lowPrio would happen twice?

Reply:
right, it would be a race basically. since both demotions end in lowPrio, the demotion would be idempotent. does this seem ok?

Reply:
My intuition would be to do the following refactor:

1) Rebrand "refunds" as "demotions"
2) Add a something like an `unsignedBid` (in addition to `signedBlindedBeaconBlock`).
3) Upsert a demotion row whenever there's a simulation failure for an optimistic builder. The upsert populates `signedBlindedBeaconBlock` in the case where the winning block's re-simulated loses the race.

Does this make sense?

Reply:
yeah, that makes a lot of sense for sure! i think this will simplify things :-) thanks. will respond on this thread again when that is done.

Reply:
ok check out this commit: https://github.com/michaelneuder/mev-boost-relay/pull/1/commits/c04edc8a190b0db8f532a521dece89865ed0ad98

here is a summary of the changes:
0) create the `TableBuilderDemotions` (fixed the typo in a follow up commit)
1) in the table add columns for UnsignedBidTrace and BlockHash.
2) add the `UpsertBuilderDemotion` function which updates the row if the signed block and validator registration are not nil (meaning a refund will need to be processed because the invalid block was published).
2) in `startOptimisticBlockProcessor()` if the block simulation fails, Upsert into that table with just the bidtrace
3) in `handleGetPayload()` if the block simulation fails, Upsert into that table with the bidtrace, signed blinded block, and the validator registation.